torch/csrc/jit/interned_strings.h

```diff
@@ -3,146 +3,252 @@
+_(onnx, Gemm) \
+_(onnx, LSTM) \
+_(onnx, Mul) \
+_(onnx, Pow) \
+_(onnx, RNN) \
+_(onnx, Sub) \
+_(onnx, Transpose) \
+/* end */
 
 // These symbols are attribute keys.  They are shared between both ONNX and ATen
-// operators (you disambiguate their meaning by looking at the operator itself)
+// operators (you disambiguate their meaning by looking at the operator itself).
+// In general, you only need to define attribute keys that are used by
+// onnx or prim; ATen attributes are automatically generated in FORALL_ATTR_BASE_SYMBOLS.
+
+#define FORALL_ATTR_EXTRA_SYMBOLS(_) \
+_(attr, Subgraph) \
+_(attr, axis) \
+_(attr, broadcast) \
+_(attr, device) \
+_(attr, inplace) \
+_(attr, is_zero) \
+_(attr, perm) \
+_(attr, sizes) \
+_(attr, transA) \
+_(attr, transB) \
+/* end */
 
 #define FORALL_ATTR_SYMBOLS(_) \
-_(Subgraph) \
-_(alpha) \
-_(axis) \
-_(broadcast) \
-_(device) \
-/* _(dim) conflicts with ATen */ \
-_(end) \
-_(exponent) \
-_(inplace) \
-_(is_zero) \
-_(keepdim) \
-_(length) \
-_(other) \
-_(perm) \
-/* _(size) conflicts with ATen */ \
-/* _(sizes) conflicts with ATen */ \
-_(start) \
-_(step) \
-_(transA) \
-_(transB) \
-_(value) \
-_(__ATTR_END)
+FORALL_ATTR_BASE_SYMBOLS(_) \
+FORALL_ATTR_EXTRA_SYMBOLS(_)
 
 #define FORALL_BUILTIN_SYMBOLS(_) \
-FORALL_JIT_SYMBOLS(_) \
-FORALL_ATEN_SYMBOLS(_) \
-FORALL_ATEN_EXTRA_SYMBOLS(_) \
-FORALL_ONNX_SYMBOLS(_) \
-FORALL_ATTR_SYMBOLS(_) \
-_(__BUILTIN_END)
-
-  enum BuiltinSymbol {
-    #define DEFINE_SYMBOL(s) \
-      k##s,
-    FORALL_BUILTIN_SYMBOLS(DEFINE_SYMBOL)
-    #undef DEFINE_SYMBOL
-    kLastSymbol, //where we start counting for new symbols
-  };
+  FORALL_ONNX_SYMBOLS(_) \
+  FORALL_ATEN_SYMBOLS(_) \
+  FORALL_ATTR_SYMBOLS(_) \
+  FORALL_PRIM_SYMBOLS(_) \
+  /* end */
+
+// Note [Symbol allocation]
+// ~~~~~~~~~~~~~~~~~~~~~~~~
+//
+//  1. Symbol namespace is split up into namespaces.  The byte structure
+//  of our symbols is TUUUUUUU, where T is the tag byte and U are the unique
+//  bytes.
+//
+//  2. We only maintain a single counter for the unique bytes, which means that
+//  we take 256 more space than we would have if we maintained multiple
+//  counters.
+//
+//  3. The first unique_start symbols are reserved for "built-in" symbols.
+//  These symbols are allocated at compile time and get put into the intern
+//  table at process startup time.  Since it's pretty easy to maintain a
+//  distinct counter for every built-in namespace, we let the unique bytes of
+//  built-in symbols to overlap (this is why unique_start is a max)
+//
+//  4. The intended access pattern for built-in symbols is onnx::MatMul
+//  in the torch::jit namespace (this is a Symbol).
+//
+// The code here is not very economical but it gets the job done.
+
 
+// Built-in constant definition strategy:
+// - Enum is the most convenient way to generate a contiguous sequence
+//   of numbers for an identifier.
+// - However, an enum gives you a fresh type which is implicitly convertible
+//   integers.
```

Comment:
There's also `class enum` that doesn't have this problem, right?

Reply:
Actually, we are using "enum class". It is not the implicit convertibility that is the problem, it is the fresh type, and we want our builtins to be Symbol. Also, the number allocated by the enum class is not the one of the symbol (which needs to be bit-OR'ed with the namespace mask)